Perform symbol publication using Microsoft.SymbolUploader.Build.Task

dougbu opened the pull request:
- aspnet/AspNetCore-Internal#2126
- add new stage template and project that publishes symbols to MSDL and SymWeb

dougbu committed:
Perform symbol publication using Microsoft.SymbolUploader.Build.Task
- aspnet/AspNetCore-Internal#2126
- add new stage template and project that publishes symbols to MSDL and SymWeb
  - install `dotnet` version matching global.json content
  - restore and build the new project

dougbu commented on .azure/pipelines/stages/publish-symbols.yml:
```diff
@@ -0,0 +1,67 @@
+      installNodeJs: false
+      isTestingJob: false
+      jobDisplayName: ${{ parameters.jobDisplayName }}
+      jobName: '_'
```
`jobName` is not useful here because no other jobs can depend on this one.

natemcmaster replied:
This affects the UI as well. Can we set it to something meaningful?

dougbu replied:
`jobName` doesn't matter when `jobDisplayName` is set

dougbu commented on .azure/pipelines/stages/publish-symbols.yml:
```diff
@@ -0,0 +1,67 @@
+            --no-restore
+            "/p:SymbolDirectory=$(Build.StagingDirectory)/symbols/"
+            "/p:MicrosoftSymbolServerPAT=$(microsoft-symbol-server-pat)"
+            "/p:SymwebSymbolServerPAT=$(symweb-symbol-server-pat)"
```
I've linked our official build pipeline to the right variable variable group to get these PATs.

dougbu commented on .azure/pipelines/stages/publish-symbols.yml:
```diff
@@ -0,0 +1,67 @@
+# publish-symbols.yml
```
Once we have eng/common/CheckSymbols.ps1, adding a step to confirm the uploads succeeded will be trivial.

natemcmaster replied:
It's there now. Just merged https://github.com/aspnet/AspNetCore/pull/10405

dougbu commented on .azure/pipelines/ci.yml:
```diff
@@ -17,383 +17,401 @@ pr:
-      displayName: Run eng/scripts/CodeCheck.ps1
+stages:
+- stage: Build
+  displayName: ' '
```
Something in AzDO uses this prefix for every job in the main stage when the "Multi-stage pipelines" preview feature is off. Making the display name non-`null` and non-empty avoids that ugly repetition.

natemcmaster replied:
On the other hand, this now causes the new UI to display nothing for the stage name. I think we should encourage the new UI and avoid workarounds like this for the old UI...which is going away relatively soon.

![image](https://user-images.githubusercontent.com/2696087/58042256-07508280-7aef-11e9-9449-589d35aceba5.png)

dougbu reviewed (commented):
https://dnceng.visualstudio.com/internal/_build/results?buildId=193044 is the build to watch because the new stage won't exist in PR verification.

dougbu commented:
@joeloff the official build of the same branch (modulo irrelevant whitespace in the YAML) is still failing the Sign Check step with the same errors as described in #10241. The same thing is happening in all other AspNetCore-ci-official builds, including the one for 'master' after you merged #10340 (see https://dnceng.visualstudio.com/internal/_build/results?buildId=192996&view=logs&j=f31c9f97-4411-58e7-49ac-fc73f645e6b6&t=ef92295d-dfa0-5952-f0c2-498c3b48582b&l=128 and https://dnceng.visualstudio.com/internal/_build/results?buildId=192996&view=logs&j=f31c9f97-4411-58e7-49ac-fc73f645e6b6&t=ef92295d-dfa0-5952-f0c2-498c3b48582b&l=168). Any ideas?

My guess is that using `$(TargetDir)` for the `@(_cabs)` item group is incorrect and should be ``$(SharedFramework.TargetDir)`. In the build for your commit, the `$(TargetDir)` is `F:\workspace\_work\1\s\artifacts\bin\SharedFrameworkLib\x64\Release\` and `$(SharedFramework.TargetDir)` is `F:\workspace\_work\1\s\artifacts\bin\SharedFramework\x64\Release\en-US\`. I'll send a PR and do a trial "official" build to see if I'm right&hellip;

dougbu commented on eng/PublishSymbols.proj:
```diff
@@ -0,0 +1,59 @@
+    <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
+    <ManifestsPath>artifacts\manifests\</ManifestsPath>
+    <DisablePackageReferenceRestrictions>true</DisablePackageReferenceRestrictions>
+  </PropertyGroup>
```
Could add a `$(SymbolExpirationInDays)` default if we want. Far as I can see, no value is passed to the symbol servers unless there's a `SymbolExpirationInDays` pipeline or queue time variable set i.e. something in the environment when this project is built. Thoughts?

natemcmaster replied:
Let's match what arcade does, so set to 3650. https://github.com/dotnet/arcade/blob/fc9dd270f84398306ec28d723f0c4d032e3f2803/eng/common/PublishToSymbolServers.proj#L38

dougbu commented:
- this PR also addresses #10113
- new trial official build is https://dnceng.visualstudio.com/internal/_build/results?buildId=193229

dougbu commented:
New trial official build is https://dnceng.visualstudio.com/internal/_build/results?buildId=193490

natemcmaster commented on .azure/pipelines/stages/publish-symbols.yml:
```diff
@@ -0,0 +1,80 @@
+  stageName: ''
+
+stages:
+- stage: ${{ parameters.stageName }}
```
```suggestion
- stage: Deploy
```

As this template isn't designed to be re-usable, can we simplify the number of parameters and move these values inline?

dougbu committed:
Address PR comments
- expire symbols after 10 years (or so)
- switch from stages to updating our release pipeline
  - revert changes to ci.yml

nits:
- add newlines to logging of published symbols
- add `$env:DOTNET_SKIP_FIRST_TIME_EXPERIENCE` to ci.yml

dougbu commented:
:up::date: to match what's been running fine in https://dnceng.visualstudio.com/internal/_releaseProgress?_a=release-pipeline-progress&releaseId=7988 and many previous iterations of the release pipeline. (Remaining issues are limited to excluding packages that contain assemblies lacking symbols.)

This branch, modulo a rebase and newlines in the symbol packages upload list, also built successfully internally: https://dnceng.visualstudio.com/internal/_build/results?buildId=197523

And, BTW I've addressed your comments @natemcmaster

dougbu commented:
Another Helix timeout at 4 hours. Most of the tests I looked at timed out though the failures on Debian.8.Amd64.Open were due to `System.Security.Cryptography.CryptographicException`s. In any case, nothing conceivably related to the changes in this PR.

Merging&hellip;
